File: packages/lynx-sqlite/ios/SqliteModule.swift
import Foundation
import Lynx
import SQLite3

private let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)
/// Integral doubles within ±2^53 bind as INTEGER so id comparisons are exact.
private let MAX_SAFE_INTEGER = 9007199254740992.0

private struct SqliteError: Error {
    let message: String
    init(_ message: String) { self.message = message }
}

/// Process-wide registry of open databases keyed by handle. Handles outlive
/// any single LynxView (Lynx may instantiate the module class per view —
/// same pattern as WebSocketTaskStore). Each entry owns a serial queue: all
/// statements for one database run serially off the JS thread, which is
/// what makes the JS-side operation queue's transaction guarantees hold.
final class SqliteStore {

    static let shared = SqliteStore()

    final class Entry {
        let db: OpaquePointer
        let queue: DispatchQueue
        init(db: OpaquePointer, queue: DispatchQueue) {
            self.db = db
            self.queue = queue
        }
    }

    private let lock = NSLock()
    private var nextHandle = 1
    private var entries: [Int: Entry] = [:]

    func register(db: OpaquePointer) -> Int {
        lock.lock(); defer { lock.unlock() }
        let handle = nextHandle
        nextHandle += 1
        entries[handle] = Entry(db: db, queue: DispatchQueue(label: "com.sigx.sqlite.db\(handle)"))
        return handle
    }

    func get(_ handle: Int) -> Entry? {
        lock.lock(); defer { lock.unlock() }
        return entries[handle]
    }

    func remove(_ handle: Int) -> Entry? {
        lock.lock(); defer { lock.unlock() }
        return entries.removeValue(forKey: handle)
    }
}

/// Embedded SQLite database module (system libsqlite3, no pod).
///
/// JS usage (via the `@sigx/lynx-sqlite` wrapper, not directly):
///
///   NativeModules.Sqlite.open(name, options, cb)            -> { handle }
///   NativeModules.Sqlite.execute(handle, sql, params, cb)   -> { rows, rowsAffected, insertId? }
///   NativeModules.Sqlite.executeBatch(handle, stmts, cb)    -> { rowsAffected }
///   NativeModules.Sqlite.beginTransaction/commit/rollback(handle, cb)
///   NativeModules.Sqlite.close(handle, cb)
///   NativeModules.Sqlite.deleteDatabase(name, cb)
///
/// Every callback resolves with the result dictionary on success or
/// `{ error }` on failure. All statement work runs on the handle's serial
/// queue (see `SqliteStore`) so the JS thread is never blocked and
/// statements for one database never interleave.
///
/// Files live in `Library/Application Support/SQLite/<name>` — persisted,
/// but not in the iCloud-backed Documents tree (chat history can be large).
class SqliteModule: NSObject, LynxModule {

    @objc static var name: String { "Sqlite" }

    @objc static var methodLookup: [String: String] {
        [
            "open": NSStringFromSelector(#selector(open(_:options:callback:))),
            "execute": NSStringFromSelector(#selector(execute(_:sql:params:callback:))),
            "executeBatch": NSStringFromSelector(#selector(executeBatch(_:statements:callback:))),
            "beginTransaction": NSStringFromSelector(#selector(beginTransaction(_:callback:))),
            "commit": NSStringFromSelector(#selector(commit(_:callback:))),
            "rollback": NSStringFromSelector(#selector(rollback(_:callback:))),
            "close": NSStringFromSelector(#selector(close(_:callback:))),
            "deleteDatabase": NSStringFromSelector(#selector(deleteDatabase(_:callback:))),
        ]
    }

    required override init() { super.init() }
    required init(param: Any) { super.init() }

    // MARK: - JS-callable methods

    @objc func open(_ name: String?, options: NSDictionary?, callback: LynxCallbackBlock?) {
        guard let name = name, Self.isPlainName(name) else {
            callback?(["error": "Database name must be a plain file name"])
            return
        }
        let path = Self.databasePath(for: name)
        try? FileManager.default.createDirectory(
            atPath: (path as NSString).deletingLastPathComponent,
            withIntermediateDirectories: true)
        var db: OpaquePointer?
        let flags = SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX
        guard sqlite3_open_v2(path, &db, flags, nil) == SQLITE_OK, let handle = db else {
            let message = db.map { String(cString: sqlite3_errmsg($0)) } ?? "Unable to open database"
            if let db = db { sqlite3_close_v2(db) }
            callback?(["error": message])
            return
        }
        sqlite3_exec(handle, "PRAGMA journal_mode=WAL", nil, nil, nil)
        callback?(["handle": SqliteStore.shared.register(db: handle)])
    }

    @objc func execute(_ handle: NSNumber, sql: String?, params: NSArray?, callback: LynxCallbackBlock?) {
        guard let entry = SqliteStore.shared.get(handle.intValue) else {
            callback?(["error": "No open database for handle \(handle)"])
            return
        }
        guard let sql = sql, !sql.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            callback?(["error": "SQL is required"])
            return
        }
        entry.queue.async {
            do {
                callback?(try Self.run(entry.db, sql: sql, params: params))
            } catch let e as SqliteError {
                callback?(["error": e.message])
            } catch {
                callback?(["error": error.localizedDescription])
            }
        }
    }

    @objc func executeBatch(_ handle: NSNumber, statements: NSArray?, callback: LynxCallbackBlock?) {
        guard let entry = SqliteStore.shared.get(handle.intValue) else {
            callback?(["error": "No open database for handle \(handle)"])
            return
        }
        guard let statements = statements, statements.count > 0 else {
            callback?(["rowsAffected": 0])
            return
        }
        entry.queue.async {
            do {
                try Self.exec(entry.db, "BEGIN IMMEDIATE")
                do {
                    var total = 0
                    for (i, item) in statements.enumerated() {
                        guard let stmt = item as? NSDictionary, let sql = stmt["sql"] as? String else {
                            throw SqliteError("Statement \(i) has no sql")
                        }
                        let result = try Self.run(entry.db, sql: sql, params: stmt["params"] as? NSArray)
                        total += (result["rowsAffected"] as? Int) ?? 0
                    }
                    try Self.exec(entry.db, "COMMIT")
                    callback?(["rowsAffected": total])
                } catch {
                    try? Self.exec(entry.db, "ROLLBACK")
                    throw error
                }
            } catch let e as SqliteError {
                callback?(["error": e.message])
            } catch {
                callback?(["error": error.localizedDescription])
            }
        }
    }

    @objc func beginTransaction(_ handle: NSNumber, callback: LynxCallbackBlock?) {
        runControl(handle, "BEGIN IMMEDIATE", callback)
    }

    @objc func commit(_ handle: NSNumber, callback: LynxCallbackBlock?) {
        runControl(handle, "COMMIT", callback)
    }

    @objc func rollback(_ handle: NSNumber, callback: LynxCallbackBlock?) {
        runControl(handle, "ROLLBACK", callback)
    }

    private func runControl(_ handle: NSNumber, _ sql: String, _ callback: LynxCallbackBlock?) {
        guard let entry = SqliteStore.shared.get(handle.intValue) else {
            callback?(["error": "No open database for handle \(handle)"])
            return
        }
        entry.queue.async {
            do {
                try Self.exec(entry.db, sql)
                callback?([String: Any]())
            } catch let e as SqliteError {
                callback?(["error": e.message])
            } catch {
                callback?(["error": error.localizedDescription])
            }
        }
    }

    @objc func close(_ handle: NSNumber, callback: LynxCallbackBlock?) {
        guard let entry = SqliteStore.shared.remove(handle.intValue) else {
            callback?(["error": "No open database for handle \(handle)"])
            return
        }
        entry.queue.async {
            if sqlite3_close_v2(entry.db) == SQLITE_OK {
                callback?([String: Any]())
            } else {
                callback?(["error": String(cString: sqlite3_errmsg(entry.db))])
            }
        }
    }

    @objc func deleteDatabase(_ name: String?, callback: LynxCallbackBlock?) {
        guard let name = name, Self.isPlainName(name) else {
            callback?(["error": "Database name must be a plain file name"])
            return
        }
        let path = Self.databasePath(for: name)
        let fm = FileManager.default
        for suffix in ["", "-wal", "-shm"] {
            let p = path + suffix
            if fm.fileExists(atPath: p) {
                do {
                    try fm.removeItem(atPath: p)
                } catch {
                    callback?(["error": error.localizedDescription])
                    return
                }
            }
        }
        callback?([String: Any]())
    }

    // MARK: - statement execution

    /// The name becomes a filesystem path — restrict it to a plain file
    /// name so a caller bypassing the JS wrapper can't traverse out of the
    /// SQLite directory. Mirrors the JS-side validation.
    private static let plainName = try! NSRegularExpression(pattern: "^[A-Za-z0-9._-]+$")
    private static func isPlainName(_ name: String) -> Bool {
        if name == "." || name == ".." { return false }
        let range = NSRange(name.startIndex..., in: name)
        return plainName.firstMatch(in: name, options: [], range: range) != nil
    }

    private static func databasePath(for name: String) -> String {
        let base = NSSearchPathForDirectoriesInDomains(
            .applicationSupportDirectory, .userDomainMask, true).first ?? NSTemporaryDirectory()
        return ((base as NSString).appendingPathComponent("SQLite") as NSString)
            .appendingPathComponent(name)
    }

    private static func exec(_ db: OpaquePointer, _ sql: String) throws {
        if sqlite3_exec(db, sql, nil, nil, nil) != SQLITE_OK {
            throw SqliteError(String(cString: sqlite3_errmsg(db)))
        }
    }

    private static func run(_ db: OpaquePointer, sql: String, params: NSArray?) throws -> [String: Any] {
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK, let stmt = stmt else {
            throw SqliteError(String(cString: sqlite3_errmsg(db)))
        }
        defer { sqlite3_finalize(stmt) }
        try bind(stmt, params: params, db: db)

        var rows: [[String: Any]] = []
        loop: while true {
            switch sqlite3_step(stmt) {
            case SQLITE_ROW:
                rows.append(try readRow(stmt))
            case SQLITE_DONE:
                break loop
            default:
                throw SqliteError(String(cString: sqlite3_errmsg(db)))
            }
        }

        let readOnly = sqlite3_stmt_readonly(stmt) != 0
        var result: [String: Any] = [
            "rows": rows,
            "rowsAffected": readOnly ? 0 : Int(sqlite3_changes(db)),
        ]
        let head = firstKeyword(sql)
        if !readOnly && (head == "INSERT" || head == "REPLACE") {
            result["insertId"] = sqlite3_last_insert_rowid(db)
        }
        return result
    }

    private static func readRow(_ stmt: OpaquePointer) throws -> [String: Any] {
        var row: [String: Any] = [:]
        for i in 0..<sqlite3_column_count(stmt) {
            let column = String(cString: sqlite3_column_name(stmt, i))
            switch sqlite3_column_type(stmt, i) {
            case SQLITE_NULL:
                row[column] = NSNull()
            case SQLITE_INTEGER:
                row[column] = sqlite3_column_int64(stmt, i)
            case SQLITE_FLOAT:
                row[column] = sqlite3_column_double(stmt, i)
            case SQLITE_TEXT:
                row[column] = String(cString: sqlite3_column_text(stmt, i))
            default:
                throw SqliteError(
                    "BLOB columns are not supported (column \"\(column)\") — " +
                    "store a file path or base64 TEXT instead")
            }
        }
        return row
    }

    /// Typed 1-based binds. The JS side pre-coerces params to string | number | null.
    private static func bind(_ stmt: OpaquePointer, params: NSArray?, db: OpaquePointer) throws {
        guard let params = params else { return }
        for (i, value) in params.enumerated() {
            let index = Int32(i + 1)
            let rc: Int32
            if value is NSNull {
                rc = sqlite3_bind_null(stmt, index)
            } else if let text = value as? String {
                rc = sqlite3_bind_text(stmt, index, text, -1, SQLITE_TRANSIENT)
            } else if let number = value as? NSNumber {
                let d = number.doubleValue
                if d.truncatingRemainder(dividingBy: 1) == 0 && abs(d) <= MAX_SAFE_INTEGER {
                    rc = sqlite3_bind_int64(stmt, index, Int64(d))
                } else {
                    rc = sqlite3_bind_double(stmt, index, d)
                }
            } else {
                throw SqliteError("Unsupported parameter type at index \(i + 1) — bind string, number or null")
            }
            if rc != SQLITE_OK {
                throw SqliteError(String(cString: sqlite3_errmsg(db)))
            }
        }
    }

    private static func firstKeyword(_ sql: String) -> String {
        var rest = Substring(sql)
        while let first = rest.first, first.isWhitespace || first == "(" {
            rest = rest.dropFirst()
        }
        return String(rest.prefix(while: { $0.isLetter })).uppercased()
    }
}

File: packages/lynx-sqlite/android/com/sigx/sqlite/SqliteModule.kt
package com.sigx.sqlite

import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteCursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteProgram
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import com.lynx.react.bridge.Callback
import com.lynx.react.bridge.JavaOnlyArray
import com.lynx.react.bridge.JavaOnlyMap
import com.lynx.react.bridge.ReadableArray
import com.lynx.react.bridge.ReadableMap
import com.lynx.react.bridge.ReadableType

/**
 * Embedded SQLite database module (framework `android.database.sqlite`,
 * no bundled C library).
 *
 * JS usage (via the `@sigx/lynx-sqlite` wrapper, not directly):
 *
 *   NativeModules.Sqlite.open(name, options, cb)            -> { handle }
 *   NativeModules.Sqlite.execute(handle, sql, params, cb)   -> { rows, rowsAffected, insertId? }
 *   NativeModules.Sqlite.executeBatch(handle, stmts, cb)    -> { rowsAffected }
 *   NativeModules.Sqlite.beginTransaction/commit/rollback(handle, cb)
 *   NativeModules.Sqlite.close(handle, cb)
 *   NativeModules.Sqlite.deleteDatabase(name, cb)
 *
 * Every callback resolves with the result map on success or `{ error }` on
 * failure. All statement work runs on the handle's single-thread executor
 * (see [SqliteStore]) so the JS thread is never blocked and statements for
 * one database never interleave.
 */
class SqliteModule(context: Context) : LynxModule(context) {

    @LynxMethod
    fun open(name: String?, @Suppress("UNUSED_PARAMETER") options: ReadableMap?, callback: Callback?) {
        val dbName = name?.takeIf { isPlainName(it) }
            ?: return err(callback, "Database name must be a plain file name")
        try {
            val file = mContext.getDatabasePath(dbName)
            file.parentFile?.mkdirs()
            val db = SQLiteDatabase.openOrCreateDatabase(file, null)
            db.enableWriteAheadLogging()
            val handle = SqliteStore.register(db)
            callback?.invoke(JavaOnlyMap().apply { putInt("handle", handle) })
        } catch (e: Exception) {
            err(callback, e.message ?: "Unable to open database")
        }
    }

    @LynxMethod
    fun execute(handle: Int, sql: String?, params: ReadableArray?, callback: Callback?) {
        val entry = SqliteStore.get(handle)
            ?: return err(callback, "No open database for handle $handle")
        if (sql.isNullOrBlank()) return err(callback, "SQL is required")
        entry.executor.execute {
            try {
                callback?.invoke(runStatement(entry.db, sql, params))
            } catch (e: Exception) {
                err(callback, e.message ?: "Statement failed")
            }
        }
    }

    @LynxMethod
    fun executeBatch(handle: Int, statements: ReadableArray?, callback: Callback?) {
        val entry = SqliteStore.get(handle)
            ?: return err(callback, "No open database for handle $handle")
        if (statements == null || statements.size() == 0) {
            callback?.invoke(JavaOnlyMap().apply { putInt("rowsAffected", 0) })
            return
        }
        entry.executor.execute {
            val db = entry.db
            var began = false
            try {
                db.beginTransaction()
                began = true
                var total = 0.0
                for (i in 0 until statements.size()) {
                    val stmt = statements.getMap(i)
                        ?: throw IllegalArgumentException("Statement $i is not an object")
                    val sql = stmt.getString("sql")
                    if (sql.isNullOrBlank()) throw IllegalArgumentException("Statement $i has no sql")
                    val result = runStatement(db, sql, stmt.getArray("params"))
                    total += result.getDouble("rowsAffected")
                }
                db.setTransactionSuccessful()
                db.endTransaction()
                began = false
                callback?.invoke(JavaOnlyMap().apply { putDouble("rowsAffected", total) })
            } catch (e: Exception) {
                if (began) {
                    try { db.endTransaction() } catch (_: Exception) {}
                }
                err(callback, e.message ?: "Batch failed")
            }
        }
    }

    @LynxMethod
    fun beginTransaction(handle: Int, callback: Callback?) {
        val entry = SqliteStore.get(handle)
            ?: return err(callback, "No open database for handle $handle")
        entry.executor.execute {
            try {
                entry.db.beginTransaction()
                callback?.invoke(JavaOnlyMap())
            } catch (e: Exception) {
                err(callback, e.message ?: "beginTransaction failed")
            }
        }
    }

    @LynxMethod
    fun commit(handle: Int, callback: Callback?) {
        endTransaction(handle, commit = true, callback)
    }

    @LynxMethod
    fun rollback(handle: Int, callback: Callback?) {
        endTransaction(handle, commit = false, callback)
    }

    private fun endTransaction(handle: Int, commit: Boolean, callback: Callback?) {
        val entry = SqliteStore.get(handle)
            ?: return err(callback, "No open database for handle $handle")
        entry.executor.execute {
            try {
                if (!entry.db.inTransaction()) {
                    err(callback, "No transaction in progress")
                    return@execute
                }
                if (commit) entry.db.setTransactionSuccessful()
                entry.db.endTransaction()
                callback?.invoke(JavaOnlyMap())
            } catch (e: Exception) {
                err(callback, e.message ?: "endTransaction failed")
            }
        }
    }

    @LynxMethod
    fun close(handle: Int, callback: Callback?) {
        val entry = SqliteStore.remove(handle)
            ?: return err(callback, "No open database for handle $handle")
        entry.executor.execute {
            try {
                entry.db.close()
                callback?.invoke(JavaOnlyMap())
            } catch (e: Exception) {
                err(callback, e.message ?: "close failed")
            }
        }
        entry.executor.shutdown() // runs the queued close, then stops the thread
    }

    @LynxMethod
    fun deleteDatabase(name: String?, callback: Callback?) {
        val dbName = name?.takeIf { isPlainName(it) }
            ?: return err(callback, "Database name must be a plain file name")
        try {
            // Also removes the -wal/-shm sidecars.
            mContext.deleteDatabase(dbName)
            callback?.invoke(JavaOnlyMap())
        } catch (e: Exception) {
            err(callback, e.message ?: "deleteDatabase failed")
        }
    }

    // ── statement execution ────────────────────────────────────────────────

    private fun runStatement(db: SQLiteDatabase, sql: String, params: ReadableArray?): JavaOnlyMap {
        val head = firstKeyword(sql)
        return when {
            head == "SELECT" || head == "EXPLAIN" || head == "VALUES" -> query(db, sql, params)
            // A CTE prefix can front INSERT/UPDATE/DELETE — those must go
            // through the write path (the query API can't run DML).
            head == "WITH" && !WRITE_VERB.containsMatchIn(sql) -> query(db, sql, params)
            // PRAGMA reads return a cursor; PRAGMA assignments don't.
            head == "PRAGMA" && !sql.contains('=') -> query(db, sql, params)
            head == "PRAGMA" -> {
                db.execSQL(sql)
                emptyResult()
            }
            else -> write(db, sql, params, head)
        }
    }

    private fun query(db: SQLiteDatabase, sql: String, params: ReadableArray?): JavaOnlyMap {
        val cursor = if (params == null || params.size() == 0) {
            db.rawQuery(sql, null)
        } else {
            // rawQuery(sql, selectionArgs) binds everything as TEXT —
            // a CursorFactory lets us bind typed values instead.
            db.rawQueryWithFactory(
                { _, driver, editTable, q ->
                    bindParams(q, params)
                    SQLiteCursor(driver, editTable, q)
                },
                // editTable is annotated non-null in current SDKs; "" means
                // no editable table, which is right for a read-only query.
                sql, null, "",
            )
        }
        val rows = JavaOnlyArray()
        try {
            while (cursor.moveToNext()) {
                val row = JavaOnlyMap()
                for (i in 0 until cursor.columnCount) {
                    val column = cursor.getColumnName(i)
                    when (cursor.getType(i)) {
                        Cursor.FIELD_TYPE_NULL -> row.putNull(column)
                        Cursor.FIELD_TYPE_INTEGER -> row.putDouble(column, cursor.getLong(i).toDouble())
                        Cursor.FIELD_TYPE_FLOAT -> row.putDouble(column, cursor.getDouble(i))
                        Cursor.FIELD_TYPE_STRING -> row.putString(column, cursor.getString(i))
                        else -> throw IllegalArgumentException(
                            "BLOB columns are not supported (column \"$column\") — " +
                                "store a file path or base64 TEXT instead",
                        )
                    }
                }
                rows.pushMap(row)
            }
        } finally {
            cursor.close()
        }
        return JavaOnlyMap().apply {
            putArray("rows", rows)
            putInt("rowsAffected", 0)
        }
    }

    private fun write(db: SQLiteDatabase, sql: String, params: ReadableArray?, head: String): JavaOnlyMap {
        val statement = db.compileStatement(sql)
        try {
            if (params != null) bindParams(statement, params)
            val result = JavaOnlyMap().apply { putArray("rows", JavaOnlyArray()) }
            when (head) {
                "INSERT", "REPLACE" -> {
                    val insertId = statement.executeInsert()
                    result.putInt("rowsAffected", if (insertId == -1L) 0 else 1)
                    if (insertId != -1L) result.putDouble("insertId", insertId.toDouble())
                }
                "UPDATE", "DELETE", "WITH" -> {
                    result.putInt("rowsAffected", statement.executeUpdateDelete())
                }
                else -> {
                    // DDL and friends — no row count.
                    statement.execute()
                    result.putInt("rowsAffected", 0)
                }
            }
            return result
        } finally {
            statement.close()
        }
    }

    /** Typed 1-based binds. The JS side pre-coerces params to string | number | null. */
    private fun bindParams(program: SQLiteProgram, params: ReadableArray) {
        for (i in 0 until params.size()) {
            val index = i + 1
            when (params.getType(i)) {
                ReadableType.Null -> program.bindNull(index)
                ReadableType.String -> program.bindString(index, params.getString(i))
                ReadableType.Number -> {
                    val d = params.getDouble(i)
                    // Integral doubles bind as INTEGER so id comparisons are exact.
                    if (d == Math.floor(d) && !d.isInfinite() && Math.abs(d) <= MAX_SAFE_INTEGER) {
                        program.bindLong(index, d.toLong())
                    } else {
                        program.bindDouble(index, d)
                    }
                }
                else -> throw IllegalArgumentException(
                    "Unsupported parameter type at index $index — bind string, number or null",
                )
            }
        }
    }

    private fun firstKeyword(sql: String): String {
        val m = HEAD_KEYWORD.find(sql)
        return m?.groupValues?.get(1)?.uppercase() ?: ""
    }

    private fun emptyResult(): JavaOnlyMap = JavaOnlyMap().apply {
        putArray("rows", JavaOnlyArray())
        putInt("rowsAffected", 0)
    }

    private fun err(callback: Callback?, message: String) {
        callback?.invoke(JavaOnlyMap().apply { putString("error", message) })
    }

    /**
     * The name becomes a filesystem path — restrict it to a plain file name
     * so a caller bypassing the JS wrapper can't traverse out of the app's
     * database directory. Mirrors the JS-side validation.
     */
    private fun isPlainName(name: String): Boolean =
        name != "." && name != ".." && PLAIN_NAME.matches(name)

    private companion object {
        val PLAIN_NAME = Regex("^[A-Za-z0-9._-]+$")
        val HEAD_KEYWORD = Regex("^[\\s(]*([A-Za-z]+)")
        val WRITE_VERB = Regex("\\b(insert|update|delete|replace)\\b", RegexOption.IGNORE_CASE)
        const val MAX_SAFE_INTEGER = 9007199254740992.0 // 2^53
    }
}
